client-tools/Shell/TelemetricsHandler.cpp

```diff
@@ -0,0 +1,380 @@
+
+    basics::StringBuffer compressedBody;
+
+    auto res =
+        encoding::gzipCompress(reinterpret_cast<uint8_t const*>(body.data()),
+                               body.size(), compressedBody);
+    if (res != TRI_ERROR_NO_ERROR) {
+      break;
+    }
+    std::unordered_map<std::string, std::string> headers;
+    headers.emplace(StaticStrings::ContentTypeHeader,
+                    StaticStrings::MimeTypeJson);
+    headers.emplace(StaticStrings::ContentLength,
+                    std::to_string(compressedBody.size()));
+    headers.emplace(StaticStrings::ContentEncoding,
+                    StaticStrings::EncodingGzip);
+
+    ClientFeature& client =
+        _server.getFeature<HttpEndpointProvider, ClientFeature>();
+    std::unique_ptr<Endpoint> newEndpoint(Endpoint::clientFactory(endpoint));
+
+    std::unique_lock lk(_mtx);
+
+    TRI_ASSERT(newEndpoint.get() != nullptr);
+    std::unique_ptr<httpclient::GeneralClientConnection> connection(
+        httpclient::GeneralClientConnection::factory(
+            client.getCommFeaturePhase(), newEndpoint, 30, 60, 3, sslProtocol));
+
+    TRI_ASSERT(connection != nullptr);
+
+    if (_httpClient != nullptr) {
+      _httpClient.reset(nullptr);
+    }
+    _httpClient = std::make_unique<httpclient::SimpleHttpClient>(
+        connection.get(), params);
+    TRI_ASSERT(_httpClient != nullptr);
+
+    lk.unlock();
+
+    std::unique_ptr<httpclient::SimpleHttpResult> response(_httpClient->request(
+        rest::RequestType::POST, relative, compressedBody.data(),
+        compressedBody.size(), headers));
+    lk.lock();
+    result = this->checkHttpResponse(response, false);
+    if (result.ok()) {
+      auto returnCode = response->getHttpReturnCode();
+      if (returnCode == 200) {
+        break;
+      } else if (returnCode == 301 || returnCode == 302 || returnCode == 307) {
+        if (numRedirects < maxRedirects) {
+          bool found;
+          url = response->getHeaderField(StaticStrings::Location, found);
+          if (found) {
+            numRedirects++;
+          } else {
+            url = originalUrl;
+            numRedirects = 0;
+          }
+        } else {
+          url = originalUrl;
+          numRedirects = 0;
+        }
+      }
+    } else {
+      numRedirects = 0;
+    }
+    if (numRedirects == 0) {
+      timeoutInSecs *= 3;
+      timeoutInSecs = std::min<uint32_t>(600, timeoutInSecs);
+      _runCondition.wait_for(lk, std::chrono::seconds(timeoutInSecs),
+                             [this] { return _server.isStopping(); });
+    }
+    _httpClient.reset(nullptr);
+    connection.reset(nullptr);
+  }
+}
+
+void TelemetricsHandler::getTelemetricsInfo(VPackBuilder& builder) {
+  // as this is for printing the telemetrics result in the tests, we send the
+  // telemetrics object if the request returned ok and the error if not to
+  // parse the error in the tests
+  std::unique_lock lk(_mtx);
+  if (_telemetricsResponse.ok()) {
```

Comment:
This constructor already locks the mutex calling `_mtx.lock()`

Reply:
I can add a condition variable there that would be signaled by `fetchTelemetricsFromServer`, but then, if this function that would wait for the condition variable is called before `fetchTelemetricsFromServer` is called, it would wait for the condition to be satisfied until it timed out, which would take a few seconds, because it would not be useful to have a very small timeout